### ISSUE TITLE
Self-referencing classes need to declare themselves as immutable

### ISSUE THREAD

**Issue by danielFesenmeyer** [MutabilityDetector/MutabilityDetector#109]
Self-referencing classes are reported as not immutable per default. An example for this is a tree. A workaround is to specify an AllowedReason which declares themself immutable, which is not really intuitive.

Can be reproduced like this:
```
public final class TreeTest {

    // Throws:
    //    org.mutabilitydetector.unittesting.MutabilityAssertionError:
    //    Expected: org.mutabilitydetector.unittesting.evaluation.TreeTest$Node to be IMMUTABLE
    //    but: org.mutabilitydetector.unittesting.evaluation.TreeTest$Node is actually NOT_IMMUTABLE
    //    Reasons:
    //    Field can have collection with mutable element type (java.util.List<org.mutabilitydetector.unittesting.evaluation.TreeTest$Node>) assigned to it. [Field: children, Class: org.mutabilitydetector.unittesting.evaluation.TreeTest$Node]
    //    Allowed reasons:
    //    None.
    @Test
    public void notHelpfulFailure() {
        MutabilityAssert.assertInstancesOf(Node.class, areImmutable());
    }

    @Test
    public void workaround() {
        MutabilityAssert.assertInstancesOf(Node.class, areImmutable(), provided(Node.class).isAlsoImmutable());
    }

    public static final class Node {
        private final List<Node> children;

        public Node(final List<Node> children) {
            this.children = Collections.unmodifiableList(new ArrayList<>(children));
        }

        public List<Node> getChildren() {
            return children;
        }
    }
}
```

**Comment by Grundlefleck** [comment]
Good diagnosis. I definitely agree the test case you provided should pass without the workaround. What do you think in terms of the generics of `children` in the example you provided? Should it be limited to specifying a collection type as a precise, non-extensible class, in other words, no `? extends` or `? super`?

I'm not sure exactly where this logic should be applied. I'll need to think about the right place in code to recognise this case, without being either too restrictive or too lax.

**Comment by Grundlefleck** [comment]
Thanks to @sband this issue was fixed (#113 ). Will be in the next release.

**Comment by danielFesenmeyer** [comment]
Thanks for your great work, guys 🥇
